### PR TITLE
[Merged by Bors] - feat: port Mathlib.Data.FP.Basic

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
Remaining things:
 - [ ] `divNatLtTwoPowₓ` with or without ₓ.
 - [ ] how to handle `undefined` meta
 - [ ] next_dn_pos
 - [ ] check naming and capitalization is correct

**Comment by hrmacbeth** [comment]
I think our naming convention
https://github.com/leanprover-community/mathlib4/wiki/Porting-wiki#naming-convention
would say `FP`, not `Fp` (since it's an acronym).

File: Mathlib.lean [hunk 1]
```diff
@@ -214,6 +214,7 @@ import Mathlib.Data.Fin.Fin2
 import Mathlib.Data.Finite.Defs
 import Mathlib.Data.Finset.Basic
 import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Fp.Basic
```
Comment (mo271) [inline Mathlib.lean:217]:
```suggestion
import Mathlib.Data.FP.Basic
```

**Comment by mo271** [comment]
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60to_additive.60.20with.20.60inductive.60

**Comment by Komyyy** [comment]
At `Part.unwrap` in `Data/Part.lean`, `undefined` was ported to `lcProof`.
Indeed, it seem to work.

**Comment by mo271** [comment]
Cool, everything seems to work now. 
So for divNatLtTwoPow we go without `ₓ`, I suppose.
Is this ready for review, @Komyyy?

File: Mathlib/Data/FP/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  prec_pos : 0 < prec
+  prec_max : prec ≤ emax
+attribute [nolint docBlame] FloatCfg.prec FloatCfg.emax FloatCfg.prec_pos FloatCfg.prec_max
+#align fp.float_cfg FP.FloatCfg
+
+variable [C : FloatCfg]
+
+-- Porting note: no longer needed
+-- include C
+
```
Comment (jcommelin) [inline Mathlib/Data/FP/Basic.lean:46]:
```suggestion
```
This is not something that we'll need to return to later, so it's fine to just remove it.

File: Mathlib/Data/FP/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+/- warning: fp.div_nat_lt_two_pow -> FP.divNatLtTwoPow is a dubious translation:
+lean 3 declaration is
+  forall [C : FP.FloatCfg], Nat -> Nat -> Int -> Bool
+but is expected to have type
+  Nat -> Nat -> Int -> Bool. -/
+@[nolint docBlame]
+def divNatLtTwoPow (n d : ℕ) : ℤ → Bool
+  | Int.ofNat e => n < d.shiftl e
+  | Int.negSucc e => n.shiftl e.succ < d
+#align fp.div_nat_lt_two_pow FP.divNatLtTwoPow
```
Comment (jcommelin) [inline Mathlib/Data/FP/Basic.lean:154]:
```suggestion
@[nolint docBlame]
def divNatLtTwoPow (n d : ℕ) : ℤ → Bool
  | Int.ofNat e => n < d.shiftl e
  | Int.negSucc e => n.shiftl e.succ < d
#align fp.div_nat_lt_two_pow FP.divNatLtTwoPowₓ -- Porting note: TC argument `[C : FP.FloatCfg]` no longer present
```

File: Mathlib/Data/FP/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+inductive RMode
+  | NE -- round to nearest even
+  deriving Inhabited
+#align fp.rmode FP.RMode
+
+@[nolint docBlame]
+class FloatCfg where
+  (prec emax : ℕ)
+  PrecPos : 0 < prec
+  PrecMax : prec ≤ emax
```
Comment (jcommelin) [inline Mathlib/Data/FP/Basic.lean:38]: These should in fact start with lowercase. Only types (and props) start with a capital letter.
```suggestion
  precPos : 0 < prec
  precMax : prec ≤ emax
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**